rfcs/20200211-tf-types.md

```diff
@@ -0,0 +1,147 @@
+
+### General Principles
+This module should not contain any implementation code. An advantage of that is that users exploring the implementation of specific types will not need to inspect this module. However, users who do not wish to inspect the code may visit the documentation of these generic types to better understand specifically what are the concrete subclasses of this type expected to do.
+
+The `tf.types` module may depend on external packages (such as `numpy`) _strictly for the purpose of defining type annotations and documentation_. No dependencies to other TensorFlow interfaces are allowed. Any dependencies on external packages which themselves depend on TensorFlow are expressly forbidden.
+
+Changes definitions inside `tf.types` must be approved by TensorFlow leads, and typically should be accompanied by an RFC.
+
+All type declarations are compatible with [pytype](https://github.com/google/pytype).
+
+It is recommended that internal and external type annotations, `isinstance` and `issubclass` checks use these types, eventually deprecating helpers like `tf.is_tensor`. However, concrete types continue to exist - for example, variables are instances of `tf.Variable`, which is now a subclass of `tf.types.Variable`.
+
+Class type definitions define a minimum of abstract methods and properties which are required for pytype compatibility.
+
+### Support for `tf.function`'s `input_signature`
+The type system listed here can be expanded to allow input signatures using type annotations, see for instance [this thread](https://github.com/tensorflow/tensorflow/issues/31579).
+
+### Initial Type Hierarchy
+TensorFlow generally adopts an incremental development method. This RFC aims to remain consistent with that.
+
+Below are listed the major types presently used in TensorFlow. All types included in this list are subject to [normal compatibility rules](https://www.tensorflow.org/guide/versions), so they are unlikely to change in the future. It is therefore preferable to maintain a strict minimum of orthogonal declarations and carefully vet any additions.
+
+Most of these symbols will not be initially exported as public symbols. Only internal submodules will be able to use unexported types. The unexported types may be gradually exposed under `tf.types` or under `tf.types.experimental`.
+
+The initial type hierarchy is focused on V2 symbols. We expect to encounter places where these symbols would not be compatible with V1 code; in such cases, the V1 symbols will not be affected.
+
+#### Types created by this RFC
+
+These types will be added with the initial creation of the `tf.types` namespace.
+
+* Core tensor types
+
+  * `DType`
+  * `Shape`
+  * `Tensor` - generic dense tensor
+
+    * `Symbol` - the regular graph tensor
+    * `Value` - eager tensors
+
+  * `TensorLike` - any type that can be implicitly converted to `Tensor` (see for example https://github.com/tensorflow/addons/blob/master/tensorflow_addons/utils/types.py)
```

Comment:
How will `TensorLike` interact with custom ->`Tensor` conversion machinery? If a user registers a new conversion function, should they also [register](https://docs.python.org/3/library/abc.html#abc.ABCMeta.register) a "virtual" subclass of `TensorLike`?

Reply:
Good question. Since using ABC may have performance implications as you pointed out, we may need to think of a different mechanism, like a concrete `TensorLike` superclass (instead of a `Union` generic) that also defines `to_tensor` or defining a dunder method similar to `__ndarray__`.

Reply:
I think that regardless of the implementation, the relationship between `TensorLike`, `tf.convert_to_tensor` and `tf.register_tensor_conversion_function` should be specified to avoid confusion. 

Personally, I'm fine with `TensorLike` being non-extensible, i.e. if you want your custom type to be `TensorLike`, convert it to such explicitly.

Reply:
My initial expectation is that `isinstance(x, tf.types.TensorLike)` would work for anything registered with `tf.register_tensor_conversion_function`. I think we cannot do that while requiring `TensorLike` to be a protocol (e.g. np.ndarray doesn't define to_tensor, but can be converted to one).

Reply:
Added a section about this - PTAL.

It turns out that protocols do support static type checking, and can support isinstance in newer Python versions too.

I'm not sure that `register_tensor_conversion_function` can be made compatible with static type checking. `ABCMeta.register` might be the equivalent of it, but it lacks the means to specify the conversion logic if you don't control the class (e.g. how do you specify the conversion logic for  `list` with ABCMeta.register?). And besides, if we replace `register_tensor_conversion_function`, we might as well opt for the leaner protocols (`hasattr` seems slightly faster than a bare `isinstance`). So my preference would be to use protocol as alternative to  `register_tensor_conversion_function`, and eventually deprecate it.